### PR TITLE
Handle subexpressions, parentheses and binary operators

### LINKED ISSUE #82: Add support for binary operators like +,-,*,/ along with precedence rules

### PR DESCRIPTION
**Pull request by jbcoe**

### CODE REVIEW

**Comment by philipcraig** [comment]
Doesn't seem to have settled down ready for review yet? When merged I would change it to be `Result` without the `panic`s but that can wait.

**Comment by philipcraig** [comment]
At least fix the clippy/lint issues?

**Comment by jbcoe** [comment]
This PR feels a bit ugly - I think we can do better.

File: src/parser.rs [hunk 1]
```diff
@@ -727,9 +825,46 @@ mod tests {
...
+    }
+
+    parse_expression_matcher_test_case! {
+        name: nested_parenthesised_identifier,
+        input: "(((x)));",
+        matcher: match_identifier!("x".to_string()),
+    }
+
+    parse_expression_matcher_test_case! {
+        name: parenthesised_chained_binary_operators,
```
Comment (philipcraig) [inline src/parser.rs:840]:
```suggestion
        name: parenthesised_chained_binary_operators_parenthesise_left,
```

File: src/parser.rs [hunk 2]
```diff
@@ -727,9 +825,46 @@ mod tests {
...
+                match_identifier!("y".to_string()),
+                OpName::Plus
+            ),
+            match_identifier!("z".to_string()),
+            OpName::Plus
+        ),
+    }
+
+    parse_expression_matcher_test_case! {
+        name: parenthesised_chained_binary_operators_2,
```
Comment (philipcraig) [inline src/parser.rs:854]:
```suggestion
        name: parenthesised_chained_binary_operators_parenthesise_right,
```

File: src/parser.rs [hunk 3]
```diff
@@ -742,4 +877,52 @@ mod tests {
...
+                match_identifier!("a".to_string()),
+                match_identifier!("b".to_string()),
+                OpName::Multiply
+            ),
+            match_binary_op!(
+                match_identifier!("c".to_string()),
+                match_identifier!("d".to_string()),
+                OpName::Divide
+            ),
+            OpName::Plus
```
Comment (philipcraig) [inline src/parser.rs:925]: I like the tests. Very clear

File: src/parser.rs [hunk 4]
```diff
@@ -154,59 +153,158 @@ impl<'a> Parser<'a> {
...
-        left: Expression,
-        operation: OpName,
+        mut expressions: VecDeque<Expression>,
+        mut operators: VecDeque<OpName>,
     ) -> Option<Expression> {
-        self.read_token();
-        self.parse_expression().map_or_else(
-            || {
+        if expressions.len() != operators.len() + 1 {
+            panic!(
```
Comment (philipcraig) [inline src/parser.rs:180]: I can clean up these `panic`s into `Err`/`Result` after merge if you like

**Review by philipcraig — APPROVED** [review]
I know you have concerns -- it's passing tests that we want to pass. Merge it and let's have at refactoring it.

**Comment by ajbennieston** [comment]
I like this as a starting point; having some code that works, and some tests we can expand on, is better than having an idea and no code.

I think I'd have approached the "bit in the middle" (finding precedence of operators) differently, but this approach feels simpler. My idea was to build a tree and parse left to right. If a higher precedence operator appears, we replace the closest `rhs` with the new node `op(lhs, rhs)` where `lhs` is the old `rhs`...

The advantage of having code in and working, with tests, is that we can explore alternate parsing strategies with some guarantee that they are at least as good as the one they're replacing, through the safety net of the tests.

That is, to parse
```
1 + 3 * 5
```
would produce something like:
```
(lhs=1)
(op=+, lhs=1)
(op=+, lhs=1, rhs=3)
(op=*, lhs=3)
(op=*, rhs=5)
(op=+, lhs=1, rhs=(op=*, lhs=3, rhs=5))
```

File: src/parser.rs (hunk 4 above)
Comment (jbcoe) [reply src/parser.rs:180]: That would be awesome, thanks!

**Pull request merged**